nemo/collections/tts/models/ssl_tts.py

```diff
@@ -0,0 +1,524 @@
+        sched_backbone_config = optim_backbone_config.pop("sched", None)
+        OmegaConf.set_struct(optim_backbone_config, True)
+
+        OmegaConf.set_struct(optim_downstream_config, False)
+        sched_downstream_config = optim_downstream_config.pop("sched", None)
+        OmegaConf.set_struct(optim_downstream_config, True)
+
+        optim_backbone = instantiate(optim_backbone_config, params=self.encoder.parameters(),)
+        optim_downstream = instantiate(
+            optim_downstream_config,
+            params=itertools.chain(
+                self.downstream_nets.parameters(),
+                self.sv_linear.parameters(),
+                self.content_linear.parameters(),
+                self.sv_loss.parameters(),
+            ),
+        )
+
+        if sched_backbone_config is not None and sched_downstream_config is not None:
+
+            scheduler_backbone = WarmupPolicy(
+                optimizer=optim_backbone,
+                max_steps=None,
+                min_lr=sched_backbone_config.min_lr,
+                warmup_steps=sched_backbone_config.warmup_steps,
+            )  # Use warmup to delay start
+            sch1_dict = {
+                'scheduler': scheduler_backbone,
+                'interval': 'step',
+            }
+
+            scheduler_downstream = WarmupPolicy(
+                optimizer=optim_downstream,
+                max_steps=None,
+                min_lr=sched_downstream_config.min_lr,
+                warmup_steps=sched_downstream_config.warmup_steps,
+            )
+            sch2_dict = {
+                'scheduler': scheduler_downstream,
+                'interval': 'step',
+            }
+
+            return [optim_backbone, optim_downstream], [sch1_dict, sch2_dict]
+        else:
+            return [optim_backbone, optim_downstream]
+
+    def get_mel_spectrogram(self, audio, audio_len):
+        EPSILON = 1e-9
+        window_fn = torch.hann_window
+
+        stft_cfg = self._cfg.preprocessor
+        spec = torch.stft(
+            input=audio,
+            n_fft=stft_cfg.n_fft,
+            hop_length=stft_cfg.n_window_stride,
+            win_length=stft_cfg.n_window_size,
+            window=window_fn(stft_cfg.n_window_size, periodic=False).float().to(self.device) if window_fn else None,
+            return_complex=True,
+            center=True,
+        )
+
+        if spec.dtype in [torch.cfloat, torch.cdouble]:
+            spec = torch.view_as_real(spec)
+        spec = torch.sqrt(spec.pow(2).sum(-1) + EPSILON)
+
+        mel = torch.matmul(self.fb.to(spec.dtype), spec)
+        log_mel = torch.log(torch.clamp(mel, min=torch.finfo(mel.dtype).tiny))
+        mel_len = torch.ceil(audio_len / stft_cfg.n_window_stride).long()
+        log_mel_normalized = features.normalize_batch(log_mel, mel_len, "per_feature")
+
+        return log_mel_normalized, mel_len
+
+    def _forward(self, input_signal=None, input_signal_length=None, for_export=False, normalize_content=True):
```

Comment:
May want to add a note that this exists to be called both from `forward()` and `forward_for_export()`.

Reply:
added comment